File: src/infrastructure/ProbabilisticTaker.cpp
#include "ProbabilisticTaker.hpp"
#include <iostream>
#include <iomanip>

namespace omm::infrastructure {

ProbabilisticTaker::ProbabilisticTaker(
    std::shared_ptr<events::EventBus> bus,
    double       trade_probability,
    unsigned int seed)
    : bus_(std::move(bus))
    , trade_probability_(trade_probability)
    , rng_(seed)                              // 固定种子，确保每次仿真结果一致
    , dist_(0.0, 1.0)                         // [0, 1) 均匀分布
    , side_dist_(0, 1) {}                     // {0, 1} 用于随机选择买/卖方向

void ProbabilisticTaker::register_handlers() {
    // 模式：观察者 — 将本组件注册为 QuoteGeneratedEvent 的订阅者
    bus_->subscribe<events::QuoteGeneratedEvent>(
        [this](const events::QuoteGeneratedEvent& evt) {
            this->on_quote_generated(evt);
        }
    );
}

void ProbabilisticTaker::on_quote_generated(
    const events::QuoteGeneratedEvent& event) {

    // 独立掷骰子：以 trade_probability_ 的概率决定是否触发成交
    double roll = dist_(rng_);
    if (roll >= trade_probability_) {
        // 本次不成交，静默跳过
        return;
    }

    // 随机选择成交方向：
    //   0 → Hit Bid  → 客户卖出（Side::Sell）→ 我们买入，持仓增加
    //   1 → Lift Ask → 客户买入（Side::Buy）  → 我们卖出，持仓减少
    bool hit_bid = (side_dist_(rng_) == 0);
    events::Side side  = hit_bid ? events::Side::Sell : events::Side::Buy;
    double       price = hit_bid ? event.bid_price     : event.ask_price;

    // 构造成交事件（TradeExecutedEvent）
    events::TradeExecutedEvent trade{
        event.instrument_id, // 合约 ID 继承自报价事件
        side,
        price,
        1,                   // 固定成交量：1 手（MVP 简化）
        event.timestamp
    };

    std::cout << "[成交模拟] "
              << event.instrument_id
              << "  " << (hit_bid ? "Hit Bid" : "Lift Ask")
              << " @ $" << std::fixed << std::setprecision(4) << price
              << "  (掷骰结果: " << std::setprecision(2) << roll
              << " < " << trade_probability_ << ")\n";

    // 发布成交事件，触发 PositionManager 和 DeltaHedger 的处理逻辑
    bus_->publish(trade);
}

} // namespace omm::infrastructure

File: python/heritage/infrastructure/probabilistic_taker.py
# ============================================================
# 文件：probabilistic_taker.py
# 职责：概率成交模拟器 — 订阅报价事件，以固定概率模拟客户成交行为。
#
# 设计：
#   - 使用固定随机种子（42）确保仿真结果可重现
#   - 30% 概率触发成交；成交方向随机（Hit Bid / Lift Ask 各 50%）
#
# 事件订阅：QuoteGeneratedEvent
# 事件发布：TradeExecutedEvent
# ============================================================

from __future__ import annotations
import random

from events.event_bus import EventBus
from events.events import QuoteGeneratedEvent, TradeExecutedEvent, Side


class ProbabilisticTaker:
    """模拟市场参与者（随机概率成交）。"""

    def __init__(
        self,
        bus:              EventBus,
        trade_probability: float = 0.10,  # 默认 10% 成交概率
        seed:             int   = 42,     # 固定种子，确保可重现
    ) -> None:
        self._bus              = bus
        self._trade_probability = trade_probability
        self._rng              = random.Random(seed)  # 隔离随机状态

    def register_handlers(self) -> None:
        """向 EventBus 注册报价事件处理器。"""
        self._bus.subscribe(QuoteGeneratedEvent, self._on_quote_generated)

    def _on_quote_generated(self, event: QuoteGeneratedEvent) -> None:
        """报价事件处理器：以概率 trade_probability 模拟客户成交。

        流程：
          1. 掷随机数，若 roll < trade_probability：触发成交
          2. 随机选择方向：Hit Bid（客户卖出）或 Lift Ask（客户买入）
          3. 发布 TradeExecutedEvent
        """
        roll = self._rng.random()
        if roll >= self._trade_probability:
            return  # 未触发成交

        # 随机选择成交方向
        hit_bid = self._rng.randint(0, 1) == 0
        side    = Side.Sell if hit_bid else Side.Buy
        price   = event.bid_price if hit_bid else event.ask_price

        direction = "Hit Bid" if hit_bid else "Lift Ask"
        print(
            f"[成交模拟] {event.instrument_id}  {direction} @ ${price:.4f}"
            f"  (掷骰结果: {roll:.2f} < {self._trade_probability})"
        )

        self._bus.publish(TradeExecutedEvent(
            instrument_id = event.instrument_id,
            side          = side,
            price         = price,
            quantity      = 1,
            timestamp     = event.timestamp,
        ))
